subsys/net/lib/dns/dns_sd.c

```diff
@@ -893,82 +974,209 @@ bool dns_sd_rec_match(const struct dns_sd_rec *record,
-	offs = DNS_MSG_HEADER_SIZE;
+	query += DNS_MSG_HEADER_SIZE;
+	query_size -= DNS_MSG_HEADER_SIZE;
+	offset = DNS_MSG_HEADER_SIZE;
+	dns_sd_create_wildcard_filter(record);
+	/* valid record must have non-NULL port */
+	record->port = &dns_sd_port_zero;
+
+	/* also counts labels */
+	for (i = 0, qlabels = 0; query_size > 0;) {
+		qsize = *query;
+		++offset;
+		++query;
+		--query_size;
+
+		if (qsize == 0) {
+			break;
+		}
 
-	/* Copy service label to '\0'-terminated buffer */
-	label_size = query[offs];
-	if (label_size == 0  || label_size > service_size - 1
-	    || offs + label_size > query_size) {
-		NET_DBG("could not get service");
-		return -EINVAL;
-	} else {
-		strncpy(service, &query[offs + 1], label_size);
-		service[label_size] = '\0';
-		offs += label_size + 1;
+		++qlabels;
+		if (qsize >= query_size) {
+			NET_DBG("claimed query size %zu > query buffer size %zu", qsize,
+				query_size);
+			return -EINVAL;
+		}
+
+		if (qsize > size[i] - 1) {
+			NET_DBG("qsize %zu > size[%d] - 1 %zu", qsize, i, size[i] - 1);
+			return -ENOBUFS;
+		}
+
+		if (i < N) {
+			/* only extract the label if there is storage for it */
+			memcpy(label[i], query, qsize);
+			label[i][qsize] = '\0';
+			size[i] = qsize;
+			++i;
+		}
+
+		offset += qsize;
+		query += qsize;
+		query_size -= qsize;
 	}
 
-	/* Copy proto label to '\0'-terminated buffer */
-	label_size = query[offs];
-	if (label_size == 0 || label_size > proto_size - 1
-	    || offs + label_size > query_size) {
-		NET_DBG("could not get proto for '%s...'", log_strdup(service));
-		return -EINVAL;
-	} else {
-		strncpy(proto, &query[offs + 1], label_size);
-		proto[label_size] = '\0';
-		offs += label_size + 1;
+	/* write-out the actual number of labels in 'n' */
+	for (*n = i; i < N; ++i) {
+		label[i] = NULL;
+		size[i] = 0;
 	}
 
-	/* Copy domain label to '\0'-terminated buffer */
-	label_size = query[offs];
-	if (label_size == 0 || label_size > domain_size - 1
-	    || offs + label_size > query_size) {
-		NET_DBG("could not get domain for '%s.%s...'",
-			log_strdup(service), log_strdup(proto));
+	if (qlabels < DNS_SD_MIN_LABELS) {
+		NET_DBG("too few labels in query %d, DNS_SD_MIN_LABELS: %d",
+			qlabels, DNS_SD_MIN_LABELS);
 		return -EINVAL;
-	} else {
-		strncpy(domain, &query[offs + 1], label_size);
-		domain[label_size] = '\0';
-		offs += label_size + 1;
 	}
 
-	/* Check that we have reached the DNS terminator */
-	if (query[offs] != 0) {
-		NET_DBG("ignoring request for '%s.%s.%s...'",
-			log_strdup(service),
-			log_strdup(proto),
-			log_strdup(domain));
+	if (qlabels > DNS_SD_MAX_LABELS) {
+		NET_DBG("too many labels in query %d, DNS_SD_MAX_LABELS: %d",
+			qlabels, DNS_SD_MAX_LABELS);
 		return -EINVAL;
 	}
 
-	offs++;
-	record->service = service;
-	record->proto = proto;
-	record->domain = domain;
+	if (qlabels > N) {
+		NET_DBG("too few buffers to extract query: qlabels: %d, N: %zu",
+			qlabels, N);
+		return -ENOBUFS;
+	}
+
+	switch (qlabels) {
+	case DNS_SD_MIN_LABELS:
```

Comment:
What if we receive a query with an unsupported "label" count? Shouldn't we return an error in such case (record will not be set then)? Even if that's not supported by the standard, we shall still be ready fol malformed packets.

Reply:
gah - sorry, I originally had 
```
default:
case DNS_SD_MAX_LABELS:
```
otherwise, for `qlabels < DNS_SD_MAX_LABELS`, it's an error

Reply:
Currently min is 3 and max is 4